### ISSUE TITLE
Plan parity causes workflow failure

### ISSUE THREAD

**Issue by chris3ware** [devsectop/tf-via-pr#337]
#### Describe the bug

With `plan-parity` enabled `terraform apply` fails

#### To Reproduce

Run the workflow in a pull request with `plan-parity: true`

#### Expected behavior

The workflow should succeed when a diff does not exist

#### Screenshots

<img width="974" alt="Screenshot 2024-10-22 at 21 13 05" src="https://github.com/user-attachments/assets/cbf0f353-35bb-4a77-bbce-31577ee6729b">

#### Additional context

Plan workflow:
https://github.com/3ware/gitops-2024/actions/runs/11464772519

Apply workflow
https://github.com/3ware/gitops-2024/actions/runs/11466493459

**Comment by rdhar** [comment]
At the outset, it's worth mentioning that `plan-parity` has a fairly niche use case and isn't warranted most times. I'll amend the docs to reflect the specific use case (within a `merge_queue` triggered workflow).

It also adds a significant time cost since it re-runs a plan before the apply which, again, isn't necessary since Terraform/Tofu will automatically handle by default if there's been any configuration drift since the plan file was generated.

**Comment by chris3ware** [comment]
I've tried to use this with a `merge_queue` and the same thing happens. I think you are correct that `plan_parity` isn't really required though because terraform reports that the plan is stale if you run `apply original-tfplan`  and something has changed.

**Comment by rdhar** [comment]
Spot-on, the lack of `plan-parity` is an internal blocker for us as well, so we've fallen back on default TF behaviour in the interim. 

Should note, this feature's desirable function is only apparent within a repo where there's multiple PRs which need to be merged frequently. In that context, each PR is blocked by the others since a successful merge/apply immediately invalidates all other PR branches' plan files as stale (TF stores an incrementing revision number in the state file and plan file for stale comparison). As a result, each PR branch needs to be re-updated to re-create its plan file afresh before merge. 

Understandable as to why TF enforces this behaviour, but it's rather unpleasantly slow and tedious to execute for each branch every time, especially when they're small changes which don't conflict or overlap with each other. 

With `plan-parity` enabled, at the point where TF apply would run, a new TF plan is run immediately before it. This newly-generated plan file is compared with the original plan file which was about to be applied. For a given PR branch which is merged with/out latest changes from main:

- If there's any significant diff between the plan files (i.e., different resources are modified), then the original plan file is passed on to TF apply where it'll error out as stale, as expected. This is because the plan file is well and truly out of sync with original intent and needs a fresh review/approval again. 
- If there's no significant diff between the plan files (i.e., the same resources are modified the same way), then the new plan file is used in place of the original one. This ensures the stale error is circumvented while the original intent of the plan file is still applied.

Scale this up for a `merge_queue` with 4 PR branches added to the queue (not uncommon after Dependabot PR updates), then TF-via-PR will take care of updating each plan file in turn and only applying those which match the originally approved intent, without any manual intervention or waiting between PRs. We're probably not the first ones to come up with this, but the novel approach has made small and frequent IaC changes much less error-prone, especially when the model is handed off to self-service teams who may not grasp the intricacies of TF just yet.

... Trying to summarise all of ^that within a line or two in the Readme is the biggest challenge 😅. Though, we're drafting a blog post on this exact topic, in case you're interested as an early-reviewer?
